File: Typescript/Day 3/task2.ts
/// <reference path="../../node_modules/@types\\node/fs.d.ts" />

import * as fs  from 'fs';

const text = fs.readFileSync('Input/Day 3.txt', 'utf8').toString();
const lines = [...text.split('\r\n')];

function getValidLine(lines: string[], comparison: number) {
  let epsilon = 0;
  let gamma = 0;
    for (let i = 0; i < lines[0].length; i++) {
    epsilon = gamma = 0;
    lines.map((line) => {
      epsilon += line[i] == '1' ? 1 : 0;
      gamma += line[i] == '0' ? 1 : 0;
    });

    lines = lines.filter(l => l[i] == (epsilon >= gamma && comparison === 0 || epsilon < gamma && comparison === 1 ? '1' : '0') );
    if (lines.length === 1)
      break;
  }
  return lines[0];
}

let oxygenGeneratorRating = getValidLine([...lines], 0);
let co2ScrubberRating = getValidLine([...lines], 1);

console.log(parseInt(oxygenGeneratorRating, 2) * parseInt(co2ScrubberRating, 2))

File: Python/Day 3/task2.py
file = open("C:\\Source\\amortsell\\AdventOfCode2021\\Input\\Day 3.txt", "r")
lines = file.readlines()

def getValidLines(lines, comparison):
  i = 0 
  ones = 0
  zeros = 0
  tmpLines = []
  while i < len(lines[0]) - 1 and len(lines) > 1:
    tmpLines = [];
    ones = 0
    zeros = 0
    for line in lines:
      if line[i] == '0':
        zeros += 1
      else:
        ones += 1

    filter = ''
    if ones >= zeros and comparison == 0 or ones < zeros and comparison == 1: 
      filter = '1'
    else: 
      filter = '0'

    for line in lines:
      if line[i] == filter:
        tmpLines.append(line)
    lines = tmpLines.copy()
    i += 1

  if lines[0][len(lines[0]) - 1] == '\n':
    lines[0] = lines[0][:-1]
  return lines[0]

oxygenGeneratorRating = int(getValidLines(lines, 0), 2)
co2ScrubberRating = int(getValidLines(lines, 1), 2)
print(oxygenGeneratorRating * co2ScrubberRating)
